### ISSUE TITLE
Exceptions thrown due to lack of SecurityManager

### ISSUE THREAD

**Issue by johnoliver** [eclipse-vertx/vertx-auth#32]
The following exception is thrown when logging in to the example app at https://github.com/vert-x3/vertx-examples/blob/master/web-examples/src/main/java/io/vertx/example/web/auth/Server.java

Currently the exception is swallowed as the DefaultSubjectContext that should log the error receives a NOPLogger, not sure why, may be due to Shiro using slf4j and the app being configured for JUL. Certainly in my project in which I have logback configured I see these exceptions, and in the demo app if you breakpoint inside DefaultSubjectContext.resolveSecurityManager you can see the exception get thrown.

```
org.apache.shiro.UnavailableSecurityManagerException: No SecurityManager accessible to the calling code, either bound to the org.apache.shiro.util.ThreadContext or as a vm static singleton.  This is an invalid application configuration.
        at org.apache.shiro.SecurityUtils.getSecurityManager(SecurityUtils.java:123)
        at org.apache.shiro.subject.support.DefaultSubjectContext.resolveSecurityManager(DefaultSubjectContext.java:106)
        at org.apache.shiro.mgt.DefaultSecurityManager.ensureSecurityManager(DefaultSecurityManager.java:411)
        at org.apache.shiro.mgt.DefaultSecurityManager.createSubject(DefaultSecurityManager.java:333)
        at org.apache.shiro.mgt.DefaultSecurityManager.createSubject(DefaultSecurityManager.java:183)
        at org.apache.shiro.mgt.DefaultSecurityManager.login(DefaultSecurityManager.java:283)
        at org.apache.shiro.subject.support.DelegatingSubject.login(DelegatingSubject.java:256)
        at io.vertx.ext.auth.shiro.impl.ShiroAuthProviderImpl.lambda$authenticate$3(ShiroAuthProviderImpl.java:79)
        at io.vertx.ext.auth.shiro.impl.ShiroAuthProviderImpl$$Lambda$31/2109966893.handle(Unknown Source)
        at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$14(ContextImpl.java:279)
        at io.vertx.core.impl.ContextImpl$$Lambda$32/1669819723.run(Unknown Source)
        at io.vertx.core.impl.OrderedExecutorFactory$OrderedExecutor.lambda$new$161(OrderedExecutorFactory.java:91)
        at io.vertx.core.impl.OrderedExecutorFactory$OrderedExecutor$$Lambda$3/1309238149.run(Unknown Source)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
```

**Comment by before30** [comment]
I got same exception when I using shiro as basic auth provider

```
JsonObject config = new JsonObject().put("properties_path", "classpath:auth.properties");
        ShiroAuth authProvider = ShiroAuth.create(vertx, ShiroAuthRealmType.PROPERTIES, config);


        AuthHandler basicAuthHandler = BasicAuthHandler.create(authProvider);
        router.route("/protected/*").handler(basicAuthHandler);
        router.route("/protected/sysinfo").handler(new SystemInfoRouteHandler());
```

```
org.apache.shiro.UnavailableSecurityManagerException: No SecurityManager accessible to the calling code, either bound to the org.apache.shiro.util.ThreadContext or as a vm static singleton.  This is an invalid application configuration.
    at org.apache.shiro.SecurityUtils.getSecurityManager(SecurityUtils.java:123) ~[shiro-core-1.2.4.jar:1.2.4]
    at org.apache.shiro.subject.support.DefaultSubjectContext.resolveSecurityManager(DefaultSubjectContext.java:106) ~[shiro-core-1.2.4.jar:1.2.4]
    at org.apache.shiro.mgt.DefaultSecurityManager.ensureSecurityManager(DefaultSecurityManager.java:411) [shiro-core-1.2.4.jar:1.2.4]
    at org.apache.shiro.mgt.DefaultSecurityManager.createSubject(DefaultSecurityManager.java:333) [shiro-core-1.2.4.jar:1.2.4]
    at org.apache.shiro.mgt.DefaultSecurityManager.createSubject(DefaultSecurityManager.java:183) [shiro-core-1.2.4.jar:1.2.4]
    at org.apache.shiro.mgt.DefaultSecurityManager.login(DefaultSecurityManager.java:283) [shiro-core-1.2.4.jar:1.2.4]
    at org.apache.shiro.subject.support.DelegatingSubject.login(DelegatingSubject.java:256) [shiro-core-1.2.4.jar:1.2.4]
    at io.vertx.ext.auth.shiro.impl.ShiroAuthProviderImpl.lambda$authenticate$3(ShiroAuthProviderImpl.java:79) [vertx-auth-shiro-3.1.0.jar:na]
    at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$15(ContextImpl.java:296) ~[vertx-core-3.1.0.jar:na]
    at io.vertx.core.impl.OrderedExecutorFactory$OrderedExecutor.lambda$new$261(OrderedExecutorFactory.java:91) ~[vertx-core-3.1.0.jar:na]
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) ~[na:1.8.0_60]
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) ~[na:1.8.0_60]
    at java.lang.Thread.run(Thread.java:745) ~[na:1.8.0_60]

```

**Comment by pmlopes** [comment]
I'm trying to reproduce this issue with 3.2.0 and by running the example above mentioned I don't get the exception. Is this issue still valid? can you still reproduce it? what are the steps?

**Comment by cescoffier** [comment]
Be aware the shiro heavily relies on thread locals to store objects. I think to reproduce it, you would need to server the request from a thread not having the thread local set (another event loop or maybe a worker).

BTW, we should check how shiro uses thread locals, as it might be an issue when the thread is shared between several clients.

**Comment by cinterloper** [comment]
i have a project that experiences this at the moment, is there any additional instrumentation that could produce more information about why this is happening?

**Comment by dazraf** [comment]
I've just come across this issue myself. It never has happened in a 'classic' vertx project. The project that it fails in uses another netty-based project called Corda (https://www.corda.net). I wonder if this is related. How does vertx ensure that the threadpool threads all received the appropriate TLS?

**Comment by dazraf** [comment]
Ok, so I've got a work-around for my project. Sharing it here in case it helps. My use-case started three Corda nodes *in-memory* for the purposes of an unit/integration test. Each nodes creates a vertx instance, deploys a verticle, which during `start()` initialises the shiro `AuthProvider`. I reconfigured the startup to start each node in a separate process and the problem disappeared! I still don't understand it and I'm digging through the Shiro code to understand it better. Also, I came across this, which may provide some clues: https://stackoverflow.com/questions/16352940/is-an-unbound-securitymanager-really-an-invalid-application-configuration-in-shi
